<div style="background:linear-gradient(145deg,#111827 0%,#1F2937 55%,#172554 100%); padding:30px; border-radius:22px; border:1px solid #475569; box-shadow:0 12px 35px rgba(0,0,0,0.35);">


<div style="display:flex; justify-content:space-between; align-items:center;">

<div style="font-size:13px; color:#94A3B8; letter-spacing:2px;">
◈ BINX AI / MODEL LAB
</div>

<div style="background:#172554; border:1px solid #3B82F6; padding:5px 11px; border-radius:20px; color:#93C5FD; font-size:11px;">
● SESSION 04.02
</div>

</div>


<div style="margin-top:24px;">

<div style="font-size:13px; color:#64748B; letter-spacing:1px;">
MODEL VALIDATION SERIES
</div>

<h1 style="color:#F8FAFC; font-size:36px; margin:6px 0 0 0; line-height:1.25;">

Week 4 — Day 2

<br>

<span style="color:#60A5FA;">
Cross-Validation
</span>

</h1>

</div>


<div style="display:flex; gap:10px; margin-top:22px;">

<div style="width:10px; height:10px; border-radius:50%; background:#60A5FA; box-shadow:0 0 10px rgba(96,165,250,0.8);"></div>

<div style="width:10px; height:10px; border-radius:50%; background:#60A5FA; box-shadow:0 0 10px rgba(96,165,250,0.8);"></div>

<div style="width:10px; height:10px; border-radius:50%; background:#60A5FA; box-shadow:0 0 10px rgba(96,165,250,0.8);"></div>

<div style="width:10px; height:10px; border-radius:50%; background:#334155;"></div>

<div style="width:10px; height:10px; border-radius:50%; background:#334155;"></div>

</div>


<p style="color:#CBD5E1; font-size:15px; line-height:1.7; margin:18px 0 0 0;">

Instead of trusting one split, we test the model across multiple folds
to measure <span style="color:#93C5FD;">consistency</span>,
<span style="color:#93C5FD;">stability</span>, and
<span style="color:#93C5FD;">reliability</span>.

</p>


<div style="display:flex; gap:10px; margin-top:24px;">

<div style="flex:1; background:#0F172A; padding:13px; border-radius:12px; border:1px solid #334155;">

<div style="font-size:11px; color:#64748B;">DATASET</div>

<div style="font-size:13px; color:#E2E8F0; margin-top:5px;">
🏦 Bank Marketing
</div>

</div>


<div style="flex:1; background:#0F172A; padding:13px; border-radius:12px; border:1px solid #334155;">

<div style="font-size:11px; color:#64748B;">VALIDATION</div>

<div style="font-size:13px; color:#E2E8F0; margin-top:5px;">
5-Fold • Stratified
</div>

</div>


<div style="flex:1; background:#0F172A; padding:13px; border-radius:12px; border:1px solid #334155;">

<div style="font-size:11px; color:#64748B;">METRIC</div>

<div style="font-size:13px; color:#E2E8F0; margin-top:5px;">
F1 Score
</div>

</div>

</div>


</div>

<div style="background:#0F172A; border-left:6px solid #60A5FA; padding:20px; border-radius:12px;">

<h2 style="color:#93C5FD; margin:0;">
🏦 Dataset Overview
</h2>

<p style="color:#CBD5E1; line-height:1.7; margin-bottom:0;">

The <b style="color:#F8FAFC;">Bank Marketing Campaigns</b> dataset contains information about direct marketing campaigns conducted by a Portuguese bank.

The campaigns mainly used phone calls to contact clients and promote a <b style="color:#60A5FA;">term deposit</b>.

The dataset contains <b style="color:#F8FAFC;">41,188 observations</b> and <b style="color:#F8FAFC;">21 columns</b>.

</p>

</div>

<div style="background:#0F172A; border-left:6px solid #22C55E; padding:20px; border-radius:12px;">

<h2 style="color:#86EFAC; margin:0;">
🎯 Dataset Goal
</h2>

<p style="color:#CBD5E1; line-height:1.7; margin-bottom:0;">

The main goal is to predict whether a client will subscribe to a <b style="color:#F8FAFC;">term deposit</b> after a bank marketing campaign.

This makes the dataset a <b style="color:#86EFAC;">binary classification problem</b>.

</p>

</div>

<div style="background:#0F172A; border-left:6px solid #38BDF8; padding:20px; border-radius:12px;">

<h2 style="color:#7DD3FC; margin:0;">
🧩 Dataset Structure
</h2>

<p style="color:#CBD5E1; line-height:1.7; margin-bottom:0;">

The dataset contains information from several categories:

<br><br>

👤 <b style="color:#F8FAFC;">Client Information</b><br>
Age, job, marital status, education, and financial status.

<br><br>

📞 <b style="color:#F8FAFC;">Contact Information</b><br>
Contact method, month, day of the week, and campaign information.

<br><br>

📊 <b style="color:#F8FAFC;">Previous Campaign Information</b><br>
Previous contacts and the outcome of previous campaigns.

<br><br>

🌍 <b style="color:#F8FAFC;">Social & Economic Information</b><br>
Employment, consumer price, confidence, and interest-rate indicators.

</p>

</div>

<div style="background:#0F172A; border-left:6px solid #F59E0B; padding:20px; border-radius:12px;">

<h2 style="color:#FBBF24; margin:0;">
🔍 Features
</h2>

<p style="color:#CBD5E1; line-height:1.7; margin-bottom:0;">

The dataset includes a mixture of <b style="color:#FBBF24;">numerical</b> and <b style="color:#FBBF24;">categorical</b> features.

<br><br>

<b style="color:#F8FAFC;">Numerical Features:</b><br>
age, campaign, pdays, previous, emp.var.rate, cons.price.idx, cons.conf.idx, euribor3m, nr.employed

<br><br>

<b style="color:#F8FAFC;">Categorical Features:</b><br>
job, marital, education, default, housing, loan, contact, month, day_of_week, poutcome

</p>

</div>

<div style="background:#0F172A; border-left:6px solid #22C55E; padding:20px; border-radius:12px;">

<h2 style="color:#86EFAC; margin:0;">
🎯 Target Variable — y
</h2>

<p style="color:#CBD5E1; line-height:1.7; margin-bottom:0;">

The target variable is <b style="color:#F8FAFC;">y</b>.

It indicates whether the client subscribed to a term deposit:

<br><br>

🟢 <b style="color:#86EFAC;">yes</b> → The client subscribed to a term deposit.

<br>

🔴 <b style="color:#FCA5A5;">no</b> → The client did not subscribe.

</p>

</div>

<div style="background:#1C1917; border-left:6px solid #F59E0B; padding:20px; border-radius:12px;">

<h2 style="color:#FBBF24; margin:0;">
⚠️ Important Feature — duration
</h2>

<p style="color:#D6D3D1; line-height:1.7; margin-bottom:0;">

The <b style="color:#FBBF24;">duration</b> feature represents the duration of the last phone call in seconds.

Although this feature can strongly improve prediction performance, its value is only known <b style="color:#FBBF24;">after the call has taken place</b>.

Therefore, we will exclude <b style="color:#FBBF24;">duration</b> from our predictive features to keep the modeling scenario realistic.

</p>

</div>

<div style="background:linear-gradient(135deg,#0F172A,#172554); padding:22px; border-radius:14px; border:1px solid #3B82F6;">

<h2 style="color:#93C5FD; margin-top:0;">
🔄 Why Is This Dataset Suitable for Cross-Validation?
</h2>

<p style="color:#CBD5E1; line-height:1.7; margin-bottom:0;">

Because this is a <b style="color:#F8FAFC;">binary classification</b> problem with an imbalanced target, a single train/test split may give us a misleading estimate of model performance.

Cross-validation allows us to evaluate the model across multiple folds and obtain a more stable estimate.

For this reason, we will use <b style="color:#60A5FA;">5-Fold Stratified Cross-Validation</b> and evaluate the model using the <b style="color:#60A5FA;">F1-score</b>.

</p>

</div>

<div style="background:linear-gradient(145deg,#111827 0%,#1F2937 55%,#172554 100%); padding:22px; border-radius:18px; border:1px solid #475569; box-shadow:0 8px 25px rgba(0,0,0,0.25);">

<div style="font-size:12px; color:#60A5FA; letter-spacing:2px; margin-bottom:8px;">
STEP 01 / DATA INITIALIZATION
</div>

<h2 style="color:#F8FAFC; margin:0; font-size:25px;">
📂 Load & Inspect the Dataset
</h2>

<p style="color:#CBD5E1; font-size:14px; line-height:1.7; margin-bottom:0; margin-top:12px;">

Before applying Cross-Validation, we first load the Bank Marketing dataset
and perform a quick inspection of its structure, dimensions, columns,
and sample records.

</p>

<div style="margin-top:15px; padding:12px 15px; background:#0F172A; border-left:4px solid #60A5FA; border-radius:8px; color:#94A3B8; font-size:13px;">

🎯 <b style="color:#93C5FD;">Goal:</b>
Understand the data we are going to validate before building the model.

</div>

</div>

In [12]:
import pandas as pd

df = pd.read_csv("bank-additional-full.csv", sep=";")

print("Dataset loaded successfully!")
print(f"Shape: {df.shape}")

Dataset loaded successfully!
Shape: (41188, 21)


In [13]:
df.head().style \
    .set_properties(**{
        'background-color': '#0F172A',
        'color': '#E2E8F0',
        'border-color': '#334155',
        'text-align': 'left'
    }) \
    .set_table_styles([
        {
            'selector': 'th',
            'props': [
                ('background-color', '#172554'),
                ('color', '#93C5FD'),
                ('font-weight', 'bold'),
                ('border-color', '#3B82F6')
            ]
        }
    ])

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,261,1,999,0,nonexistent,1.100000,93.994000,-36.400000,4.857000,5191.000000,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,149,1,999,0,nonexistent,1.100000,93.994000,-36.400000,4.857000,5191.000000,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,226,1,999,0,nonexistent,1.100000,93.994000,-36.400000,4.857000,5191.000000,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,151,1,999,0,nonexistent,1.100000,93.994000,-36.400000,4.857000,5191.000000,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,307,1,999,0,nonexistent,1.100000,93.994000,-36.400000,4.857000,5191.000000,no


In [14]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  str    
 2   marital         41188 non-null  str    
 3   education       41188 non-null  str    
 4   default         41188 non-null  str    
 5   housing         41188 non-null  str    
 6   loan            41188 non-null  str    
 7   contact         41188 non-null  str    
 8   month           41188 non-null  str    
 9   day_of_week     41188 non-null  str    
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  str    
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null  float64
 1

In [15]:
from IPython.display import display, HTML

columns_html = "".join(
    f"""
    <span style="
        display:inline-block;
        background:#172554;
        color:#93C5FD;
        border:1px solid #334155;
        padding:6px 10px;
        margin:4px;
        border-radius:8px;
        font-size:12px;
    ">{col}</span>
    """
    for col in df.columns
)

display(HTML(f"""
<div style="
    background:#0F172A;
    padding:15px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="
        color:#64748B;
        font-size:11px;
        letter-spacing:1.5px;
        margin-bottom:8px;
    ">
        DATASET FEATURES
    </div>
    
    <div>
        {columns_html}
    </div>
</div>
"""))

<div style="background:linear-gradient(145deg,#111827 0%,#1F2937 55%,#172554 100%); padding:22px; border-radius:18px; border:1px solid #475569; box-shadow:0 8px 25px rgba(0,0,0,0.25);">

<div style="font-size:12px; color:#60A5FA; letter-spacing:2px; margin-bottom:8px;">
02.01 / MODEL VALIDATION
</div>

<h2 style="color:#F8FAFC; margin:0; font-size:27px;">
🔄 What Cross-Validation Does
</h2>

<div style="height:3px; width:65px; background:#60A5FA; margin-top:14px; border-radius:5px;"></div>

</div>

<div style="background:#0F172A; border:1px solid #334155; padding:22px; border-radius:14px; margin-top:12px;">

<h3 style="color:#93C5FD; margin-top:0;">
🔄 Why Do We Need Cross-Validation?
</h3>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8;">

A single train/test split can sometimes give us a
<b style="color:#F8FAFC;">lucky or unlucky estimate</b> of model performance.

If the test data happens to be easy, the model may appear better than it really is.
If the test data is difficult, the model may appear worse than it actually is.

<b style="color:#60A5FA;">Cross-Validation</b> reduces this problem by evaluating the model
on multiple different validation folds instead of relying on only one split.

</p>

</div>

<div style="background:linear-gradient(135deg,#0F172A,#172554); padding:22px; border-radius:14px; border:1px solid #334155;">

<h3 style="color:#60A5FA; margin-top:0;">
🧠 The Main Idea
</h3>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8;">

In <b style="color:#F8FAFC;">k-fold cross-validation</b>, the training data is divided into
<b style="color:#60A5FA;">k equal folds</b>.

The model is trained multiple times. In each round, one fold is used for validation,
while the remaining folds are used for training.

The process is repeated until every fold has been used as the validation set exactly once.

Finally, the scores from all folds are combined to produce a more
<b style="color:#60A5FA;">stable estimate</b> of the model's performance.

</p>

</div>

<div style="background:#0F172A; padding:20px; border-radius:14px; border:1px solid #334155;">

<div style="color:#94A3B8; font-size:12px; letter-spacing:1px; margin-bottom:12px;">
CROSS-VALIDATION FLOW
</div>

<div style="display:flex; gap:7px; flex-wrap:wrap;">

<div style="background:#172554; color:#93C5FD; padding:10px 15px; border-radius:8px; border:1px solid #3B82F6;">
Fold 1
</div>

<div style="background:#172554; color:#93C5FD; padding:10px 15px; border-radius:8px; border:1px solid #3B82F6;">
Fold 2
</div>

<div style="background:linear-gradient(135deg,#1D4ED8,#2563EB); color:#FFFFFF; padding:10px 15px; border-radius:8px; border:1px solid #60A5FA;">
Fold 3
</div>

<div style="background:#172554; color:#93C5FD; padding:10px 15px; border-radius:8px; border:1px solid #3B82F6;">
Fold 4
</div>

<div style="background:#172554; color:#93C5FD; padding:10px 15px; border-radius:8px; border:1px solid #3B82F6;">
Fold 5
</div>

</div>

<p style="color:#94A3B8; font-size:13px; margin-bottom:0; margin-top:14px;">

Each fold gets a turn as the validation set while the other folds are used for training.

</p>

</div>

<div style="background:linear-gradient(145deg,#111827 0%,#1F2937 55%,#172554 100%); padding:22px; border-radius:18px; border:1px solid #475569; box-shadow:0 8px 25px rgba(0,0,0,0.25);">

<div style="font-size:12px; color:#60A5FA; letter-spacing:2px; margin-bottom:8px;">
02.02 / FOLD ROTATION
</div>

<h2 style="color:#F8FAFC; margin:0; font-size:27px;">
🔁 How k-Fold Works (k = 5)
</h2>

<div style="height:3px; width:65px; background:#60A5FA; margin-top:14px; border-radius:5px;"></div>

</div>

<div style="background:#0F172A; padding:22px; border-radius:14px; border:1px solid #334155;">

<h3 style="color:#93C5FD; margin-top:0;">
🔄 Five Validation Rounds
</h3>

<table style="width:100%; border-collapse:collapse; color:#CBD5E1; font-size:14px; text-align:center;">

<tr style="background:#172554; color:#93C5FD;">
<th style="padding:12px; border-bottom:1px solid #3B82F6;">Round</th>
<th style="padding:12px; border-bottom:1px solid #3B82F6;">Training Folds</th>
<th style="padding:12px; border-bottom:1px solid #3B82F6;">Validation Fold</th>
</tr>

<tr>
<td style="padding:11px; border-bottom:1px solid #1E293B;">1</td>
<td style="padding:11px;">2, 3, 4, 5</td>
<td style="padding:11px; color:#60A5FA; font-weight:bold;">Fold 1</td>
</tr>

<tr>
<td style="padding:11px; border-bottom:1px solid #1E293B;">2</td>
<td style="padding:11px;">1, 3, 4, 5</td>
<td style="padding:11px; color:#60A5FA; font-weight:bold;">Fold 2</td>
</tr>

<tr>
<td style="padding:11px; border-bottom:1px solid #1E293B;">3</td>
<td style="padding:11px;">1, 2, 4, 5</td>
<td style="padding:11px; color:#60A5FA; font-weight:bold;">Fold 3</td>
</tr>

<tr>
<td style="padding:11px; border-bottom:1px solid #1E293B;">4</td>
<td style="padding:11px;">1, 2, 3, 5</td>
<td style="padding:11px; color:#60A5FA; font-weight:bold;">Fold 4</td>
</tr>

<tr>
<td style="padding:11px;">5</td>
<td style="padding:11px;">1, 2, 3, 4</td>
<td style="padding:11px; color:#60A5FA; font-weight:bold;">Fold 5</td>
</tr>

</table>

<p style="color:#94A3B8; font-size:13px; line-height:1.7; margin-bottom:0; margin-top:16px;">

Every data point is used for validation exactly once and for training
<b style="color:#F8FAFC;">k − 1 times</b>. This means that no data is wasted
and no single split dominates the final performance estimate.

</p>

</div>

<div style="background:linear-gradient(135deg,#0F172A,#172554); padding:20px; border-radius:14px; border:1px solid #334155;">

<div style="color:#60A5FA; font-size:12px; letter-spacing:1.5px; margin-bottom:8px;">
📌 COMMON CHOICES
</div>

<p style="color:#CBD5E1; font-size:15px; line-height:1.7; margin:0;">

Common choices for <b style="color:#F8FAFC;">k</b> are
<b style="color:#60A5FA;">5</b> and <b style="color:#60A5FA;">10</b>.

A smaller value of <b style="color:#F8FAFC;">k</b> requires less computation,
while a larger value provides more validation rounds but increases the
computational cost.

For this notebook, we will use <b style="color:#60A5FA;">5-Fold Cross-Validation</b>.

</p>

</div>

<div style="background:linear-gradient(145deg,#111827 0%,#1F2937 55%,#172554 100%); padding:22px; border-radius:18px; border:1px solid #475569; box-shadow:0 8px 25px rgba(0,0,0,0.25);">

<div style="font-size:12px; color:#60A5FA; letter-spacing:2px; margin-bottom:8px;">
02.03 / CROSS-VALIDATED SCORING
</div>

<h2 style="color:#F8FAFC; margin:0; font-size:27px;">
💻 cross_val_score
</h2>

<div style="height:3px; width:65px; background:#60A5FA; margin-top:14px; border-radius:5px;"></div>

<p style="color:#CBD5E1; font-size:14px; line-height:1.7; margin-bottom:0; margin-top:15px;">
Instead of evaluating the model on a single validation split,
<strong style="color:#93C5FD;">cross_val_score</strong> evaluates it across
multiple folds and returns one performance score for each fold.
</p>

</div>

In [16]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

data = df.copy()

data = data.drop(columns=["duration"])

categorical_cols = data.select_dtypes(include=["object"]).columns

label_encoders = {}

for col in categorical_cols:
    le = LabelEncoder()
    data[col] = le.fit_transform(data[col])
    label_encoders[col] = le

X = data.drop(columns=["y"])
y = data["y"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("X_train shape:", X_train.shape)
print("y_train shape:", y_train.shape)
print("X_test shape:", X_test.shape)
print("y_test shape:", y_test.shape)

X_train shape: (32950, 19)
y_train shape: (32950,)
X_test shape: (8238, 19)
y_test shape: (8238,)


C:\Users\lenovo\AppData\Local\Temp\ipykernel_14032\2058256427.py:8: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = data.select_dtypes(include=["object"]).columns


In [17]:
from IPython.display import display, HTML

display(HTML(f"""
<div style="
    background:linear-gradient(145deg,#111827,#172554);
    padding:20px;
    border-radius:16px;
    border:1px solid #334155;
    box-shadow:0 8px 22px rgba(0,0,0,0.25);
">

<div style="
    color:#64748B;
    font-size:11px;
    letter-spacing:2px;
    margin-bottom:12px;
">
DATA SPLIT / READY
</div>

<div style="display:flex; gap:12px;">

<div style="
    flex:1;
    background:#0F172A;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:11px;">TRAINING DATA</div>
    <div style="color:#60A5FA; font-size:20px; font-weight:bold; margin-top:5px;">
        {X_train.shape[0]:,} rows
    </div>
</div>

<div style="
    flex:1;
    background:#0F172A;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:11px;">TEST DATA</div>
    <div style="color:#93C5FD; font-size:20px; font-weight:bold; margin-top:5px;">
        {X_test.shape[0]:,} rows
    </div>
</div>

<div style="
    flex:1;
    background:#0F172A;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:11px;">FEATURES</div>
    <div style="color:#E2E8F0; font-size:20px; font-weight:bold; margin-top:5px;">
        {X_train.shape[1]}
    </div>
</div>

</div>

<div style="
    margin-top:14px;
    color:#86EFAC;
    font-size:13px;
">
● Features and target are ready for Cross-Validation
</div>

</div>
"""))

In [18]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score

model = DecisionTreeClassifier(random_state=42)

cv_scores = cross_val_score(
    model,
    X_train,
    y_train,
    cv=5,
    scoring="f1"
)

cv_scores

array([0.30729834, 0.3012987 , 0.32611465, 0.32470893, 0.3073955 ])

In [19]:
print("Cross-Validation Scores:")

for i, score in enumerate(cv_scores, start=1):
    print(f"Fold {i}: {score:.4f}")

print("\nMean F1-score:", cv_scores.mean().round(4))
print("Standard Deviation:", cv_scores.std().round(4))

Cross-Validation Scores:
Fold 1: 0.3073
Fold 2: 0.3013
Fold 3: 0.3261
Fold 4: 0.3247
Fold 5: 0.3074

Mean F1-score: 0.3134
Standard Deviation: 0.0101


<div style="background:#0F172A; padding:20px; border-radius:14px; border:1px solid #334155;">

<h3 style="color:#93C5FD; margin-top:0;">
📊 Cross-Validation Result
</h3>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8; margin-bottom:0;">

The Decision Tree achieved a mean <b style="color:#60A5FA;">F1-score of 0.3134</b>
across the five folds, with a standard deviation of only
<b style="color:#60A5FA;">0.0101</b>.

The relatively small standard deviation indicates that the model's performance
remained <b style="color:#86EFAC;">consistent across the different folds</b>.

</p>

</div>

<div style="background:linear-gradient(145deg,#111827 0%,#1F2937 55%,#172554 100%); padding:22px; border-radius:18px; border:1px solid #475569; box-shadow:0 8px 25px rgba(0,0,0,0.25);">

<div style="font-size:12px; color:#60A5FA; letter-spacing:2px; margin-bottom:8px;">
02.04 / STRATIFIED VALIDATION
</div>

<h2 style="color:#F8FAFC; margin:0; font-size:27px;">
⚖️ Stratified k-Fold for Classification
</h2>

<div style="height:3px; width:65px; background:#60A5FA; margin-top:14px; border-radius:5px;"></div>

</div>

<div style="background:#0F172A; padding:22px; border-radius:14px; border:1px solid #334155;">

<h3 style="color:#93C5FD; margin-top:0;">
⚖️ Why Does Stratification Matter?
</h3>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8;">

In classification problems, especially when the target classes are
<b style="color:#F8FAFC;">imbalanced</b>, a regular k-Fold split can accidentally
create folds with different class proportions.

<br><br>

For example, if the majority class is much larger than the minority class,
one fold might contain relatively fewer positive examples than another fold.
This can make the model's evaluation less reliable.

<br><br>

<b style="color:#60A5FA;">Stratified k-Fold</b> solves this problem by preserving
approximately the same class distribution in every fold as in the original dataset.

<br><br>

For our Bank Marketing classification task, this is especially important because
the <b style="color:#F8FAFC;">"yes"</b> class is much smaller than the
<b style="color:#F8FAFC;">"no"</b> class.

</p>

<div style="
    margin-top:18px;
    padding:14px;
    background:#172554;
    border:1px solid #334155;
    border-radius:10px;
">

<div style="color:#64748B; font-size:11px; letter-spacing:1.5px;">
KEY IDEA
</div>

<div style="color:#93C5FD; font-size:14px; margin-top:6px;">
Same class balance → more reliable folds → more reliable evaluation
</div>

</div>

</div>

<div style="background:linear-gradient(135deg,#0F172A,#172554); padding:20px; border-radius:14px; border:1px solid #334155;">

<div style="color:#60A5FA; font-size:12px; letter-spacing:1.5px;">
🏦 BANK MARKETING / CLASS BALANCE
</div>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8; margin-bottom:0; margin-top:10px;">

Our target variable <b style="color:#F8FAFC;">y</b> contains two classes:
<b style="color:#FCA5A5;">no</b> and <b style="color:#86EFAC;">yes</b>.

The <b style="color:#F8FAFC;">"yes"</b> class represents clients who subscribed
to a term deposit and is considerably smaller than the <b style="color:#F8FAFC;">"no"</b> class.

Therefore, maintaining a similar class distribution across validation folds
helps ensure that each fold provides a fair representation of the classification problem.

</p>

</div>

<div style="background:linear-gradient(145deg,#111827 0%,#1F2937 55%,#172554 100%); padding:25px; border-radius:20px; border:1px solid #475569; box-shadow:0 10px 30px rgba(0,0,0,0.3);">

<div style="font-size:12px; color:#60A5FA; letter-spacing:2px; margin-bottom:8px;">
◈ PRACTICAL LAB / MODEL VALIDATION
</div>

<h1 style="color:#F8FAFC; font-size:30px; margin:0; line-height:1.3;">
🧪 Hands-On Lab
<br>
<span style="color:#60A5FA;">Cross-Validating a Model</span>
</h1>

<div style="height:3px; width:75px; background:#38BDF8; margin-top:16px; border-radius:5px;"></div>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8; margin-bottom:0; margin-top:16px;">
In this hands-on lab, we will apply cross-validation to a classification model
and evaluate how consistently it performs across different validation folds.
We will then compare the cross-validated performance with the single-split
evaluation from the previous work.
</p>

<div style="display:flex; gap:10px; margin-top:20px;">

<div style="flex:1; background:#0F172A; padding:13px; border-radius:12px; border:1px solid #334155;">
<div style="color:#64748B; font-size:10px;">MODEL</div>
<div style="color:#E2E8F0; font-size:13px; margin-top:5px;">🌳 Decision Tree</div>
</div>

<div style="flex:1; background:#0F172A; padding:13px; border-radius:12px; border:1px solid #334155;">
<div style="color:#64748B; font-size:10px;">VALIDATION</div>
<div style="color:#E2E8F0; font-size:13px; margin-top:5px;">5-Fold CV</div>
</div>

<div style="flex:1; background:#0F172A; padding:13px; border-radius:12px; border:1px solid #334155;">
<div style="color:#64748B; font-size:10px;">METRIC</div>
<div style="color:#E2E8F0; font-size:13px; margin-top:5px;">F1 Score</div>
</div>

</div>

</div>

<div style="background:#0F172A; padding:20px; border-radius:15px; border-left:5px solid #60A5FA;">

<div style="color:#60A5FA; font-size:12px; letter-spacing:2px;">
HANDS-ON / STEP 01
</div>

<h2 style="color:#F8FAFC; margin:8px 0 10px 0;">
🔄 Evaluate the Model with 5-Fold Cross-Validation
</h2>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8; margin-bottom:0;">

We will take the <b style="color:#93C5FD;">Decision Tree Classifier</b> and
evaluate it using <b style="color:#93C5FD;">5-fold cross-validation</b>.

Instead of relying on one validation split, the training data will be divided
into five folds. The model will be trained and validated five times, with a
different fold used for validation each time.

The <b style="color:#93C5FD;">F1-score</b> will be used to measure the model's
performance on each fold.

</p>

</div>

In [20]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score

model = DecisionTreeClassifier(random_state=42)

cv_scores = cross_val_score(
    model,
    X_train,
    y_train,
    cv=5,
    scoring="f1"
)

print("5-Fold F1 Scores:")
print(cv_scores)

5-Fold F1 Scores:
[0.30729834 0.3012987  0.32611465 0.32470893 0.3073955 ]


<div style="background:linear-gradient(135deg,#0F172A,#172554); padding:20px; border-radius:15px; border:1px solid #334155;">

<h3 style="color:#93C5FD; margin-top:0;">
📊 What Did We Learn?
</h3>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8;">

The model produced slightly different F1-scores across the five folds.
This is expected because each fold contains a different subset of the
training data.

<br><br>

The scores range from approximately <b style="color:#60A5FA;">0.3013</b>
to <b style="color:#60A5FA;">0.3261</b>, showing that the model's performance
does not change dramatically between folds.

<br><br>

This gives us a more reliable view of the model's performance than relying
on only one train/validation split.

</p>

<div style="margin-top:15px; padding:12px; background:#111827; border-radius:10px; border:1px solid #334155; color:#86EFAC; font-size:13px;">
● 5 folds completed successfully
</div>

</div>

<div style="background:#0F172A; padding:20px; border-radius:15px; border-left:5px solid #60A5FA;">

<div style="color:#60A5FA; font-size:12px; letter-spacing:2px;">
HANDS-ON / STEP 02
</div>

<h2 style="color:#F8FAFC; margin:8px 0 10px 0;">
📊 Calculate Mean & Standard Deviation
</h2>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8; margin-bottom:0;">

After obtaining an F1-score from each fold, we need to summarize the
cross-validation results.

The <b style="color:#93C5FD;">mean</b> represents the average performance
of the model across all five folds, while the
<b style="color:#93C5FD;">standard deviation</b> measures how much the
performance varies between folds.

</p>

</div>

In [21]:
from IPython.display import display, HTML

display(HTML(f"""
<div style="
    background:linear-gradient(145deg,#111827,#172554);
    padding:22px;
    border-radius:16px;
    border:1px solid #334155;
    box-shadow:0 8px 22px rgba(0,0,0,0.25);
">

<div style="
    color:#64748B;
    font-size:11px;
    letter-spacing:2px;
    margin-bottom:14px;
">
STEP 02 / CV PERFORMANCE SUMMARY
</div>

<div style="display:flex; gap:12px;">

<div style="
    flex:1;
    background:#0F172A;
    padding:16px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">MEAN F1-SCORE</div>
    <div style="color:#60A5FA; font-size:26px; font-weight:bold; margin-top:6px;">
        {mean_f1:.4f}
    </div>
    <div style="color:#64748B; font-size:11px; margin-top:4px;">
        Average across 5 folds
    </div>
</div>

<div style="
    flex:1;
    background:#0F172A;
    padding:16px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">STANDARD DEVIATION</div>
    <div style="color:#93C5FD; font-size:26px; font-weight:bold; margin-top:6px;">
        {std_f1:.4f}
    </div>
    <div style="color:#64748B; font-size:11px; margin-top:4px;">
        Variation between folds
    </div>
</div>

</div>

<div style="
    margin-top:15px;
    padding:12px;
    background:#172554;
    border-radius:10px;
    border:1px solid #334155;
    color:#CBD5E1;
    font-size:13px;
">
● The low standard deviation indicates relatively consistent performance
across the five validation folds.
</div>

</div>
"""))

NameError: name 'mean_f1' is not defined

<div style="background:#0F172A; padding:20px; border-radius:15px; border:1px solid #334155;">

<h3 style="color:#93C5FD; margin-top:0;">
🔎 Interpreting the Results
</h3>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8;">

The Decision Tree achieved a mean F1-score of
<b style="color:#60A5FA;">0.3134</b> across the five folds.

The standard deviation is only
<b style="color:#60A5FA;">0.0101</b>, which indicates that the F1-score
did not vary greatly from one fold to another.

<br><br>

Therefore, the model shows relatively
<b style="color:#86EFAC;">stable performance</b> across different subsets
of the training data.

However, the relatively low F1-score means that the model's overall
classification performance is not particularly strong, so stability
does not necessarily mean high predictive performance.

</p>

</div>

<div style="background:#0F172A; padding:20px; border-radius:15px; border-left:5px solid #60A5FA;">

<div style="color:#60A5FA; font-size:12px; letter-spacing:2px;">
HANDS-ON / STEP 03
</div>

<h2 style="color:#F8FAFC; margin:8px 0 10px 0;">
🔍 Cross-Validation vs. Single Split
</h2>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8; margin-bottom:0;">

In this step, we compare the model's performance from a
<b style="color:#93C5FD;">single train/test split</b> with its average
performance from <b style="color:#93C5FD;">5-fold cross-validation</b>.

This helps us understand whether the original single-split result was
representative of the model's general performance or was influenced by
one particular data split.

</p>

</div>

In [ ]:
display(HTML(f"""
<div style="
    background:linear-gradient(145deg,#111827,#172554);
    padding:22px;
    border-radius:16px;
    border:1px solid #334155;
    box-shadow:0 8px 22px rgba(0,0,0,0.25);
">

<div style="
    color:#64748B;
    font-size:11px;
    letter-spacing:2px;
">
SINGLE-SPLIT EVALUATION
</div>

<div style="
    color:#60A5FA;
    font-size:30px;
    font-weight:bold;
    margin-top:8px;
">
{single_split_f1:.4f}
</div>

<div style="
    color:#94A3B8;
    font-size:12px;
    margin-top:4px;
">
F1-score on the held-out test set
</div>

</div>
"""))

In [ ]:
display(HTML(f"""
<div style="
    background:#0F172A;
    padding:20px;
    border-radius:16px;
    border:1px solid #334155;
">

<div style="color:#64748B; font-size:11px; letter-spacing:2px;">
PERFORMANCE COMPARISON
</div>

<div style="display:flex; gap:12px; margin-top:15px;">

<div style="
    flex:1;
    background:#111827;
    padding:16px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        SINGLE SPLIT
    </div>
    <div style="color:#60A5FA; font-size:24px; font-weight:bold; margin-top:5px;">
        {single_split_f1:.4f}
    </div>
</div>

<div style="
    flex:1;
    background:#111827;
    padding:16px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        5-FOLD CV MEAN
    </div>
    <div style="color:#93C5FD; font-size:24px; font-weight:bold; margin-top:5px;">
        {mean_f1:.4f}
    </div>
</div>

<div style="
    flex:1;
    background:#111827;
    padding:16px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        DIFFERENCE
    </div>
    <div style="color:#86EFAC; font-size:24px; font-weight:bold; margin-top:5px;">
        {abs(single_split_f1 - mean_f1):.4f}
    </div>
</div>

</div>

</div>
"""))

<div style="background:#0F172A; padding:20px; border-radius:15px; border:1px solid #334155;">

<h3 style="color:#93C5FD; margin-top:0;">
🔎 Single Split vs. Cross-Validation
</h3>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8;">

The single train/test split produced an F1-score of
<b style="color:#60A5FA;">0.3311</b>, while the 5-fold cross-validation
produced a mean F1-score of <b style="color:#60A5FA;">0.3134</b>.

<br><br>

The difference between the two results is
<b style="color:#60A5FA;">0.0177</b>. The single-split score is slightly
higher, which suggests that the particular test split used in the original
evaluation was somewhat more favorable to the model than the average
validation fold.

<br><br>

Cross-validation gives us a more reliable estimate because it evaluates the
model across multiple different subsets of the training data instead of
depending on one particular split.

</p>

<div style="margin-top:15px; padding:12px; background:#172554; border-radius:10px; border:1px solid #334155; color:#93C5FD; font-size:13px;">
💡 The cross-validation mean provides a broader view of expected model performance.
</div>

</div>

<div style="background:#0F172A; padding:20px; border-radius:15px; border-left:5px solid #60A5FA;">

<div style="color:#60A5FA; font-size:12px; letter-spacing:2px;">
HANDS-ON / STEP 04
</div>

<h2 style="color:#F8FAFC; margin:8px 0 10px 0;">
⚖️ Confirm Stratified Folds
</h2>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8; margin-bottom:0;">

Because our task is a <b style="color:#93C5FD;">classification problem</b>,
we need to make sure that each validation fold maintains a similar
distribution of the target classes.

Scikit-learn automatically uses <b style="color:#93C5FD;">StratifiedKFold</b>
when <code>cross_val_score</code> receives a classifier and an integer value
such as <code>cv=5</code>.

We will explicitly create a <b style="color:#93C5FD;">StratifiedKFold</b>
object and use it for cross-validation to confirm this behavior.

</p>

</div>

In [ ]:
from sklearn.model_selection import StratifiedKFold

stratified_cv = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

print(stratified_cv)

StratifiedKFold(n_splits=5, random_state=42, shuffle=True)


In [ ]:
from IPython.display import display, HTML

display(HTML(f"""
<div style="
    background:#0F172A;
    padding:22px;
    border-radius:18px;
    border:1px solid #334155;
    box-shadow:0 8px 25px rgba(0,0,0,0.25);
">

<div style="
    color:#64748B;
    font-size:11px;
    letter-spacing:2px;
">
STRATIFIED CROSS-VALIDATION
</div>


<div style="
    color:#F8FAFC;
    font-size:22px;
    font-weight:bold;
    margin-top:6px;
">
⚖️ 5-Fold F1 Performance
</div>


<div style="
    display:flex;
    gap:10px;
    margin-top:20px;
">


<div style="
    flex:1;
    background:#111827;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        FOLD 01
    </div>
    <div style="color:#60A5FA; font-size:21px; font-weight:bold; margin-top:5px;">
        {stratified_scores[0]:.4f}
    </div>
</div>


<div style="
    flex:1;
    background:#111827;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        FOLD 02
    </div>
    <div style="color:#60A5FA; font-size:21px; font-weight:bold; margin-top:5px;">
        {stratified_scores[1]:.4f}
    </div>
</div>


<div style="
    flex:1;
    background:#111827;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        FOLD 03
    </div>
    <div style="color:#60A5FA; font-size:21px; font-weight:bold; margin-top:5px;">
        {stratified_scores[2]:.4f}
    </div>
</div>


<div style="
    flex:1;
    background:#111827;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        FOLD 04
    </div>
    <div style="color:#60A5FA; font-size:21px; font-weight:bold; margin-top:5px;">
        {stratified_scores[3]:.4f}
    </div>
</div>


<div style="
    flex:1;
    background:#111827;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        FOLD 05
    </div>
    <div style="color:#60A5FA; font-size:21px; font-weight:bold; margin-top:5px;">
        {stratified_scores[4]:.4f}
    </div>
</div>


</div>


<div style="
    display:flex;
    gap:12px;
    margin-top:14px;
">


<div style="
    flex:1;
    background:#172554;
    padding:16px;
    border-radius:12px;
    border:1px solid #3B82F6;
">
    <div style="color:#93C5FD; font-size:10px;">
        MEAN F1-SCORE
    </div>
    <div style="color:#F8FAFC; font-size:26px; font-weight:bold; margin-top:5px;">
        {mean_f1:.4f}
    </div>
</div>


<div style="
    flex:1;
    background:#111827;
    padding:16px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#94A3B8; font-size:10px;">
        STANDARD DEVIATION
    </div>
    <div style="color:#F8FAFC; font-size:26px; font-weight:bold; margin-top:5px;">
        {std_f1:.4f}
    </div>
</div>


</div>


<div style="
    margin-top:16px;
    padding-top:13px;
    border-top:1px solid #334155;
    color:#94A3B8;
    font-size:12px;
">
    ⚖️ Stratified validation preserves class proportions across folds.
</div>


</div>
"""))

NameError: name 'stratified_scores' is not defined

<div style="background:linear-gradient(145deg,#111827,#172554); padding:22px; border-radius:17px; border:1px solid #334155;">

<div style="color:#60A5FA; font-size:12px; letter-spacing:2px;">
STRATIFIED CROSS-VALIDATION / RESULTS
</div>

<h2 style="color:#F8FAFC; margin:10px 0;">
⚖️ What Do the Results Tell Us?
</h2>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8;">

The Decision Tree achieved a mean F1-score of
<b style="color:#60A5FA;">0.3205</b> using 5-fold Stratified Cross-Validation.

The standard deviation was
<b style="color:#60A5FA;">0.0114</b>, indicating that the model's performance
remained relatively consistent across the five stratified folds.

<br><br>

The folds preserved a similar proportion of the target classes, which is
especially important because the Bank Marketing dataset is an
<b style="color:#93C5FD;">imbalanced classification problem</b>.

<br><br>

Compared with the previous cross-validation mean of
<b style="color:#60A5FA;">0.3134</b>, the stratified result is very similar.
This confirms that the classification performance is reasonably consistent
when the class distribution is preserved across folds.

</p>

<div style="margin-top:16px; padding:13px; background:#0F172A; border-radius:10px; border:1px solid #334155; color:#86EFAC; font-size:13px;">
● Stratified 5-Fold Validation Completed Successfully
</div>

</div>

In [ ]:
StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

StratifiedKFold(n_splits=5, random_state=42, shuffle=True)